### ISSUE TITLE
Crash when remuxing Program Stream to Matroska container

### ISSUE THREAD

**Issue by brad0** [mbunkus/mkvtoolnix#1217]
When trying to remux a Program Stream containing MPEG2 / AC3 to Matroska container mkvtoolnix is crashing..

Command line:
```
mkvmerge --output /home/brad/VTS_01_1.mkv --language 0:und --language 1:und '(' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_1.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' ')' --language 0:und --language 1:und '(' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' ')' --language 0:und --language 1:und '(' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' ')' --language 0:und --language 1:und '(' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' ')' --language 0:und --language 1:und '(' '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' ')' --track-order 0:0,0:1,1:0,1:1,2:0,2:1,3:0,3:1,4:0,4:1
```

Output from mkvmerge:
```
mkvmerge v7.9.0 ('Birds') 64bit
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_1.VOB': Using the demultiplexer for the format 'MPEG program stream'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_1.VOB': Processing the following files as well: "VTS_01_2.VOB", "VTS_01_3.VOB", "VTS_01_4.VOB", "VTS_01_5.VOB"
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB': Using the demultiplexer for the format 'MPEG program stream'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB': Processing the following files as well: "VTS_01_3.VOB", "VTS_01_4.VOB", "VTS_01_5.VOB"
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB': Using the demultiplexer for the format 'MPEG program stream'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB': Processing the following files as well: "VTS_01_4.VOB", "VTS_01_5.VOB"
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB': Using the demultiplexer for the format 'MPEG program stream'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB': Processing the following files as well: "VTS_01_5.VOB"
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB': Using the demultiplexer for the format 'MPEG program stream'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_1.VOB' track 0: Using the output module for the format 'MPEG-1/2'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_1.VOB' track 1: Using the output module for the format 'AC3'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB' track 0: Using the output module for the format 'MPEG-1/2'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB' track 1: Using the output module for the format 'AC3'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB' track 0: Using the output module for the format 'MPEG-1/2'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB' track 1: Using the output module for the format 'AC3'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' track 0: Using the output module for the format 'MPEG-1/2'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' track 1: Using the output module for the format 'AC3'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' track 0: Using the output module for the format 'MPEG-1/2'.
'/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' track 1: Using the output module for the format 'AC3'.
The file '/home/brad/VTS_01_1.mkv' has been opened for writing.
Warning: '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_2.VOB' track 1: This AC3 track contains 675 bytes of non-AC3 data which were skipped. The audio/video synchronization may have been lost.
Warning: Found at least one B frame without second reference in a non closed GOP.
Warning: '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_3.VOB' track 1: This AC3 track contains 899 bytes of non-AC3 data which were skipped. The audio/video synchronization may have been lost.
Warning: Found at least one B frame without second reference in a non closed GOP.
Warning: '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_4.VOB' track 1: This AC3 track contains 1347 bytes of non-AC3 data which were skipped. The audio/video synchronization may have been lost.
Warning: Found at least one B frame without second reference in a non closed GOP.
Warning: '/Multimedia/Movies/Skyfall (2012) 720p/VIDEO_TS/VTS_01_5.VOB' track 1: This AC3 track contains 1571 bytes of non-AC3 data which were skipped. The audio/video synchronization may have been lost.
Warning: Found at least one B frame without second reference in a non closed GOP.
Segmentation fault (core dumped)
```

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00001b84b935dc2b in MPEGFrameRef::TryUpdate (this=0x1b8767b470d0) at src/mpegparser/M2VParser.cpp:32
32        if (frame && frame->stamped && (timecode == -1))
(gdb) bt full
#0  0x00001b84b935dc2b in MPEGFrameRef::TryUpdate (this=0x1b8767b470d0) at src/mpegparser/M2VParser.cpp:32
No locals.
#1  0x00001b84b935ea69 in M2VParser::UpdateFrame (this=0x1b8742f8ce78, frame=0x1b8767b47080) at src/mpegparser/M2VParser.cpp:246
        i = 0
        __func__ = "UpdateFrame"
#2  0x00001b84b935ee25 in M2VParser::TimestampWaitingFrames (this=0x1b8742f8ce78) at src/mpegparser/M2VParser.cpp:287
        frame = @0x1b8767b47e80: 0x1b8767b47080
        __for_range = @0x1b8742f8ce98: {<std::_Vector_base<MPEGFrame*, std::allocator<MPEGFrame*> >> = {
            _M_impl = {<std::allocator<MPEGFrame*>> = {<__gnu_cxx::new_allocator<MPEGFrame*>> = {<No data fields>}, <No data fields>}, _M_start = 0x1b8767b47e80, _M_finish = 0x1b8767b47ee0, 
              _M_end_of_storage = 0x1b8767b47f00}}, <No data fields>}
        __for_begin = {_M_current = 0x1b8767b47e80}
        __for_end = {_M_current = 0x1b8767b47ee0}
#3  0x00001b84b935ec1e in M2VParser::OrderFrame (this=0x1b8742f8ce78, frame=0x1b87b5e65c80) at src/mpegparser/M2VParser.cpp:268
        p = 0x1b87b5e65c80
#4  0x00001b84b935f374 in M2VParser::PrepareFrame (this=0x1b8742f8ce78, chunk=0x1b87988c9390, timecode=86, picHdr=...) at src/mpegparser/M2VParser.cpp:355
        outBuf = 0x1b87b5e65c80
        bCopy = false
        pData = 0x1b873d404000 ""
        dataLen = 23051
#5  0x00001b84b935f81f in M2VParser::FillQueues (this=0x1b8742f8ce78) at src/mpegparser/M2VParser.cpp:440
        myTime = 86
        chunk = 0x1b87988c9390
        picHdr = {temporalReference = 2, frameType = 1 '\001', pictureStructure = 3, repeatFirstField = 2 '\002', topFieldFirst = 0 '\000', progressive = 128 '\200'}
        done = false
#6  0x00001b84b935df5e in M2VParser::WriteData (this=0x1b8742f8ce78, data=0x1b8724959000 "\331\006]\002j\327\232\346\201T?\314\366\204", dataSize=127124) at src/mpegparser/M2VParser.cpp:92
No locals.
#7  0x00001b84b9168da4 in mpeg1_2_video_packetizer_c::process_unframed (this=0x1b8742f8c410, packet=...) at src/output/p_mpeg1_2.cpp:214
        bytes_to_add = 127124
        state = 1
        old_memory = {<std::__shared_ptr<memory_c, (__gnu_cxx::_Lock_policy)2>> = {_M_ptr = 0x1b873d379830, _M_refcount = {_M_pi = 0x1b87909ed5a0}}, <No data fields>}
        data_ptr = 0x1b8724959000 "\331\006]\002j\327\232\346\201T?\314\366\204"
        new_bytes = 127124
#8  0x00001b84b9168a28 in mpeg1_2_video_packetizer_c::process (this=0x1b8742f8c410, packet=...) at src/output/p_mpeg1_2.cpp:182
No locals.
#9  0x00001b84b9029a5d in generic_packetizer_c::process (this=0x1b8742f8c410, packet=0x1b86c97b2d00) at src/merge/generic_packetizer.h:138
No locals.
#10 0x00001b84b912d545 in mpeg_ps_reader_c::read (this=0x1b86f9255540, requested_ptzr=0x1b8742f8c410, force=false) at src/input/r_mpeg_ps.cpp:1308
        new_packet = 0x1b86c97b2d00
        packet_pos = 475150
        packet = {m_id = {id = 224, sub_id = 0}, m_pts = -1, m_dts = -1, m_length = 2025, m_full_length = 2028, m_valid = true, 
          m_buffer = {<std::__shared_ptr<memory_c, (__gnu_cxx::_Lock_policy)2>> = {_M_ptr = 0x0, _M_refcount = {_M_pi = 0x0}}, <No data fields>}}
        track = {<std::__shared_ptr<mpeg_ps_track_t, (__gnu_cxx::_Lock_policy)2>> = {_M_ptr = 0x1b86f9258200, _M_refcount = {_M_pi = 0x1b86bd244300}}, <No data fields>}
        timecode = -1
        new_id = {id = 224, sub_id = 0}
        num_queued_bytes = 17920
#11 0x00001b84b8fee880 in generic_packetizer_c::read (this=0x1b8742f8c410) at src/merge/generic_packetizer.cpp:1195
No locals.
#12 0x00001b84b8fcd530 in pull_packetizers_for_packets () at src/merge/output_control.cpp:1808
        ptzr = @0x1b875fe46c00: {status = FILE_STATUS_MOREDATA, old_status = FILE_STATUS_MOREDATA, pack = {<std::__shared_ptr<packet_t, (__gnu_cxx::_Lock_policy)2>> = {_M_ptr = 0x0, _M_refcount = {
                _M_pi = 0x0}}, <No data fields>}, packetizer = 0x1b8742f8c410, orig_packetizer = 0x1b8742f8c410, file = 0, orig_file = 0, deferred = false}
        __for_range = @0x1b84b99dc300: {<std::_Vector_base<packetizer_t, std::allocator<packetizer_t> >> = {
            _M_impl = {<std::allocator<packetizer_t>> = {<__gnu_cxx::new_allocator<packetizer_t>> = {<No data fields>}, <No data fields>}, _M_start = 0x1b875fe46c00, _M_finish = 0x1b875fe46e80, 
              _M_end_of_storage = 0x1b875fe47000}}, <No data fields>}
        __for_begin = {_M_current = 0x1b875fe46c00}
        __for_end = {_M_current = 0x1b875fe46e80}
#13 0x00001b84b8fcd95c in main_loop () at src/merge/output_control.cpp:1877
        winner = 0x1b875fe46c00
        appended_a_track = false
#14 0x00001b84b8f7331c in main (argc=50, argv=0x7f7ffffe90d8) at src/merge/mkvmerge.cpp:2540
        args = {<std::_Vector_base<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > >> = {
            _M_impl = {<std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >> = {<__gnu_cxx::new_allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >> = {<No data fields>}, <No data fields>}, _M_start = 0x1b87a059b400, _M_finish = 0x1b87a059b588, _M_end_of_storage = 0x1b87a059b600}}, <No data fields>}
        start = 1432963879851
(gdb)
```

**Comment by mbunkus** [comment]
Thanks for the report. Unfortunately the stack trace isn't really enough to track it down and fix it. I'll need the source files for which it's happening. Can you please upload them to my [FTP server](/mbunkus/mkvtoolnix/wiki/FTP-server)? Thanks.

**Comment by brad0** [comment]
Looks like I don't need all of the files. Just one of them alone (first one) is enough to trigger the issue.

**Comment by mbunkus** [comment]
I don't think so, for several reasons:

1. Even if you add only the first file mkvmerge will automatically use the other ones unless you explicitly tell it not to by putting that file in parenthesis, e.g. `mkvmerge -o out.mkv ( VTS_01_1.VOB )`.
2. The output shown above shows several warnings from files other than `VTS_01_1.VOB`, and the crash only happens after those warnings.
3. I cannot reproduce the crash with the single file you've uploaded.

**Comment by brad0** [comment]
I wasn't saying that as if I was guessing that it would still crash. It does crash with just the one file on its own.

```
mkvmerge --output /home/brad/VTS_01_1.mkv --language 0:und --language 1:und '(' /home/brad/VTS_01_1.VOB ')' --track-order 0:0,0:1
```

```
mkvmerge v7.9.0 ('Birds') 64bit
'/home/brad/VTS_01_1.VOB': Using the demultiplexer for the format 'MPEG program stream'.
'/home/brad/VTS_01_1.VOB' track 0: Using the output module for the format 'MPEG-1/2'.
'/home/brad/VTS_01_1.VOB' track 1: Using the output module for the format 'AC3'.
The file '/home/brad/VTS_01_1.mkv' has been opened for writing.
Segmentation fault (core dumped)
```

**Comment by mbunkus** [comment]
I cannot reproduce this, neither with the current git HEAD nor with release 7.9.0, not even with the command line you've just posted. Are you 100% sure you've uploaded the same file you're testing with?

Are you using self-compiled binaries or ones provided by myself?

Can you also please try the same job on a different computer? A different Linux, on Windows…?

Note that your first post shows that you've manually added all VOB files so that they're added in parallel. This is 100% not what you actually want: first, it adds the files in parallel instead of sequentially, and second mkvmerge already processes all following VOB files automatically when you add the first file.

**Comment by mbunkus** [comment]
BTW, the file you've uploaded has an SHA1 checksum of `e128351d5463263de7223320f7168a4ce0ca1079` – just to make sure we're both using the same file.

**Comment by brad0** [comment]
Yes, that is the same file.

Well, "you" don't provide binaries for OpenBSD. As of a few days ago its binaries built from our ports / packages tree.

I'll try on Windows.

Oh, ok. I thought I had to add all of the VOB files in the GUI frontend. Thanks for pointing that out.

**Comment by mbunkus** [comment]
If it's only happening on OpenBSD then I cannot (don't want to) really help you. I only support a limited number of operating systems and versions directly, and as you know OpenBSD is not among them.

I will happily accept patches that fix such issues, of course.

**Comment by mbunkus** [comment]
Closing due to unsupported platform and not being able to reproduce it locally. If you manage to reproduce it on Windows or Linux or if you have a fix then please post about it here and I'll re-open it.

**Comment by brad0** [comment]
This issue has been fixed by this commit..

https://github.com/mbunkus/mkvtoolnix/commit/780efb34fbb1ca65516f12fc3f82596c9f84a91e

**Comment by mbunkus** [comment]
Ah, I didn't remember this issue. Thanks for verifying that it works now.
